x-pack/plugins/cases/server/services/user_actions/index.ts

```diff
@@ -55,359 +41,453 @@ import {
   PUSH_CONNECTOR_ID_REFERENCE_NAME,
 } from '../../common/constants';
 import { findConnectorIdReference } from '../transform';
-import { buildFilter, combineFilters, arraysDifference } from '../../client/utils';
-import type {
-  Attributes,
-  BuilderParameters,
-  CommonArguments,
-  CreateUserAction,
-  UserActionEvent,
-  UserActionParameters,
-} from './types';
-import { BuilderFactory } from './builder_factory';
+import { buildFilter, combineFilters } from '../../client/utils';
+import type { CaseConnectorActivity, CaseConnectorFields, PushInfo, ServiceContext } from './types';
 import { defaultSortField, isCommentRequestTypeExternalReferenceSO } from '../../common/utils';
 import type { PersistableStateAttachmentTypeRegistry } from '../../attachment_framework/persistable_state_registry';
 import { injectPersistableReferencesToSO } from '../../attachment_framework/so_references';
-import type { IndexRefresh } from '../types';
-import { isAssigneesArray, isStringArray } from './type_guards';
-import type { CaseSavedObject } from '../../common/types';
-import { UserActionAuditLogger } from './audit_logger';
-import { createDeleteEvent } from './builders/delete_case';
+import { UserActionPersister } from './operations/create';
 
 export interface UserActionItem {
   attributes: CaseUserActionAttributesWithoutConnectorId;
   references: SavedObjectReference[];
 }
 
-interface PostCaseUserActionArgs extends IndexRefresh {
-  actions: UserActionEvent[];
+interface MostRecentResults {
+  mostRecent: {
+    hits: {
+      total: number;
+      hits: SavedObjectsRawDoc[];
+    };
+  };
 }
 
-interface CreateUserActionES<T> extends IndexRefresh {
-  attributes: T;
-  references: SavedObjectReference[];
+interface ConnectorActivityAggsResult {
+  references: {
+    connectors: {
+      ids: {
+        buckets: Array<{
+          key: string;
+          reverse: {
+            connectorActivity: {
+              buckets: {
+                changeConnector: MostRecentResults;
+                createCase: MostRecentResults;
+                pushInfo: MostRecentResults;
+              };
+            };
+          };
+        }>;
+      };
+    };
+  };
 }
 
-type CommonUserActionArgs = CommonArguments;
-
-interface GetUserActionItemByDifference extends CommonUserActionArgs {
-  field: string;
-  originalValue: unknown;
-  newValue: unknown;
+interface ConnectorFieldsBeforePushAggsResult {
+  references: {
+    connectors: {
+      reverse: {
+        ids: {
+          buckets: Record<string, MostRecentResults>;
+        };
+      };
+    };
+  };
 }
 
-interface TypedUserActionDiffedItems<T> extends GetUserActionItemByDifference {
-  originalValue: T[];
-  newValue: T[];
-}
+export class CaseUserActionService {
```

Comment:
This one is going to be tough to review, all of the functionality used for creating or audit logging was moved to the persister class so the changes here look horrendous 😬